h2/src/main/org/h2/expression/condition/ConditionAndOrN.java

```diff
@@ -0,0 +1,344 @@
+                if (left instanceof Comparison && right instanceof Comparison) {
+                    // try to add conditions (A=B AND B=1: add A=1)
+                    Expression added = ((Comparison) left).getAdditionalAnd(session, (Comparison) right);
+                    if (added != null) {
+                        if (this.added == null) {
+                            this.added = new ArrayList<>();
+                        }
+                        this.added.add(added.optimize(session));
+                    }
+                }
+                break;
+            case OR:
+                if (!session.getDatabase().getSettings().optimizeOr) {
+                    break;
+                }
+                Expression reduced;
+                if (left instanceof Comparison && right instanceof Comparison) {
+                    reduced = ((Comparison) left).optimizeOr(session, (Comparison) right);
+                } else if (left instanceof ConditionIn && right instanceof Comparison) {
+                    reduced = ((ConditionIn) left).getAdditional((Comparison) right);
+                } else if (right instanceof ConditionIn && left instanceof Comparison) {
+                    reduced = ((ConditionIn) right).getAdditional((Comparison) left);
+                } else if (left instanceof ConditionInConstantSet && right instanceof Comparison) {
+                    reduced = ((ConditionInConstantSet) left).getAdditional(session, (Comparison) right);
+                } else if (right instanceof ConditionInConstantSet && left instanceof Comparison) {
+                    reduced = ((ConditionInConstantSet) right).getAdditional(session, (Comparison) left);
+                } else if (left instanceof ConditionAndOr && right instanceof ConditionAndOr) {
+                    reduced = ConditionAndOr.optimizeConditionAndOr((ConditionAndOr) left, (ConditionAndOr) right);
+                } else {
+                    // TODO optimization: convert .. OR .. to UNION if the cost
+                    // is lower
+                    break;
+                }
+                if (reduced != null) {
+                    expressions.remove(i);
+                    expressions.set(i-1, reduced);
+                    continue; // because we don't want to increment, we want to compare the new pair exposed
+                }
+            }
+
+            Expression e = ConditionAndOr.optimizeConstant(session, this, andOrType, left, right);
+            if (e != this) {
+                expressions.remove(i);
+                expressions.set(i-1, e);                
+                continue; // because we don't want to increment, we want to compare the new pair exposed
+            }
+
+            if (optimizeMerge(i)) {
+                continue;
+            }
+
+            i++;
+        }
+
+        Collections.sort(expressions, COMPARE_BY_COST);
+        
+        if (expressions.size() == 1) {
+            return expressions.get(0);
```

Comment:
Conditions should use `castToBoolean(…)` to ensure `BOOLEAN` data type of optimized expression.